rust/worker/src/segment/types.rs

```diff
@@ -26,6 +33,273 @@ impl<'a> MaterializedLogRecord<'a> {
+    // None if the record exists only in the log.
+    data_record: Option<DataRecord<'a>>,
+    // If present in the record segment then it is the offset id
+    // in the record segment at which the record was found.
+    // If not present in the segment then it is the offset id
+    // at which it should be inserted.
+    offset_id: u32,
+    // Set only for the records that are being inserted for the first time
+    // in the log since data_record will be None in such cases.
+    user_id: Option<&'a str>,
+    // There can be several entries in the log for an id. This is the final
+    // operation that needs to be done on it. For e.g.
+    // If log has [Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update] then final operation is Insert.
+    // If log has [Update, Update] then final operation is Update.
+    final_operation: Operation,
+    // This is the metadata obtained by combining all the operations
+    // present in the log for this id.
+    // E.g. if has log has [Insert(a: h), Update(a: b, c: d), Update(a: e, f: g)] then this
+    // will contain (a: e, c: d, f: g). This is None if the final operation
+    // above is Delete.
+    metadata_to_be_merged: Option<Metadata>,
+    // This is the final document obtained from the last non null operation.
+    // E.g. if log has [Insert(str0), Update(str1), Update(str2), Update()] then this will contain
+    // str2. None if final operation is Delete.
+    final_document: Option<&'a str>,
+    // Similar to above, this is the final embedding obtained
+    // from the last non null operation.
+    // E.g. if log has [Insert(emb0), Update(emb1), Update(emb2), Update()]
+    // then this will contain emb2. None if final operation is Delete.
+    final_embedding: Option<&'a [f32]>,
+}
+
+impl<'a> MaterializedLogRecordV2<'a> {
+    pub(crate) fn from_data_record(data_record: DataRecord<'a>, offset_id: u32) -> Self {
+        Self {
+            data_record: Some(data_record),
+            offset_id,
+            user_id: None,
+            final_operation: Operation::Add,
+            metadata_to_be_merged: None,
+            final_document: None,
+            final_embedding: None,
+        }
+    }
+
+    pub(crate) fn from_operation_record(
+        log_record: &'a OperationRecord,
+        offset_id: u32,
+        user_id: &'a str,
+    ) -> Self {
+        let metadata = match &log_record.metadata {
+            Some(metadata) => match update_metdata_to_metdata(metadata) {
+                Ok(m) => Some(m),
+                Err(e) => panic!("Should not panic. TODO"),
+            },
+            None => None,
+        };
+
+        let document = match &log_record.document {
+            Some(doc) => Some(doc.as_str()),
+            // TODO(Sanket): Should this be an error since this is an insert operation.
```

Comment:
lets just add the errors in now? i feel like we've let a lot of todos build and its easier to do it right first time

Reply:
Also using try_from will make this explicit

Reply:
Thanks for the tip!